### DISCUSSION TITLE
How to create a dialect which support `datetime` type parameter for `String` column

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#10749] [Usage Questions]
I have a thirdparty database and it use `YYYYmmdd` format string as date.

But user querying it by `date` or `datetime`. 

I have created a dialect which use `self.colspecs[sqltypes.String] = _SupportDateString` and 

```python
class _SupportDateString(sqltypes.String):
    def _literal_processor_string(self, dialect):
        def process(value):
            # HERE CHANGES
            if isinstance(value, date):
                value = value.strftime("%Y%m%d")
            elif isinstance(value, datetime):
                value = value.strftime("%Y%m%d%H%M%S")

            value = value.replace("'", "''")

            if dialect.identifier_preparer._double_percents:
                value = value.replace("%", "%%")

            return "'%s'" % value

        return process

    def literal_processor(self, dialect):
        return self._literal_processor_string(dialect)
```

This works with literal compile but not plain query. How can I work further?

**Comment by zzzeek** [comment]
to make a type that stores strings but deals in dates on the Python side you need to use [TypeDecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#augmenting-existing-types), you shouldn't subclass "String" as that's not how the API is intended to be used.

When using TypeDecorator, you would implement these methods:

https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.process_literal_param

https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.process_bind_param

https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.process_result_value


see the [examples](https://docs.sqlalchemy.org/en/20/core/custom_types.html#typedecorator-recipes)
